### PR TITLE
META-39: Set up GraphQL client using urql

### PR DESCRIPTION
**Pull request by cnk1dev**
## Description of issue
Set up the GraphQL client urql to fetch data from the server + added graphql-codegen
## How I implemented
- Installed [next-urql](https://github.com/urql-graphql/urql/tree/main/packages/next-urql) package
- Set up urql client using withUrqlClient hook from next-urql
- Added [graphql-codegen](https://github.com/dotansimha/graphql-code-generator) package
- Created graphql/ folder for queries
- Added graphql-codegen config & script that automatically creates types and urql query hooks in genereated/gql.ts for all queries in the graphql/ folder
- To use a query, import the corresponding hook from generated/gql.ts (see pages/example.tsx for reference)
## Other
- Added generated/gql.ts to .eslintignore because it breaks no-use-before-define rule which can't be fixed automatically with eslint --fix (the file is automatically formatted using prettier after each generation though)

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -16,13 +16,22 @@ module.exports = {
...
+        jsx: 'never',
+        ts: 'never',
+        tsx: 'never'
+      }
+    ]
   },
   env: {
     browser: true,
-    node: false
+    node: true
```
Comment (pczern) [inline .eslintrc.js:34]: let me know why `node: true`

File: .eslintrc.js [hunk 2]
```diff
@@ -16,13 +16,22 @@ module.exports = {
   ],
   plugins: ['@typescript-eslint', 'react', 'react-hooks', 'sonarjs', 'jsx-a11y'],
   parser: '@typescript-eslint/parser',
-
   rules: {
-    'react/jsx-filename-extension': [1, { extensions: ['.tsx', '.jsx'] }]
+    'react/jsx-filename-extension': [1, { extensions: ['.tsx', '.jsx'] }],
+    'import/extensions': [
+      'error',
```
Comment (pczern) [inline .eslintrc.js:22]: have a look at Arnos PR, there it is done a little different

File: .eslintrc.js [hunk 3]
```diff
@@ -16,13 +16,22 @@ module.exports = {
   ],
   plugins: ['@typescript-eslint', 'react', 'react-hooks', 'sonarjs', 'jsx-a11y'],
   parser: '@typescript-eslint/parser',
-
   rules: {
-    'react/jsx-filename-extension': [1, { extensions: ['.tsx', '.jsx'] }]
+    'react/jsx-filename-extension': [1, { extensions: ['.tsx', '.jsx'] }],
```
Comment (pczern) [inline .eslintrc.js:20]: we should always use warning level 2

File: src/graphql/repositories.graphql [hunk 4]
```diff
@@ -0,0 +1,7 @@
+query GitRepositories {
```
Comment (pczern) [inline src/graphql/repositories.graphql:1]: we should call the file `gitRepositories` if it contains one query

**Comment by pczern** [comment]
https://dev.to/xcanchal/automatically-generate-typescript-types-for-your-graphql-api-1fah article to understand GraphQL codegen

File: README.md [hunk 5]
```diff
@@ -1,30 +1,70 @@
...
+#### `prettier:all`
 
-`npm run prepare` installs the precommit hook, that will run eslint and prettier before you commit anything.
+Same as `prettier:check` but automatically fixes problems.
 
-`npm run dev` open [http://localhost:3000](http://localhost:3000) with your browser to see the result.
+#### `react:update-react-imports`
 
-Start working on your commits by adding files or modifying existing ones like `pages/index.tsx`. The pages auto-update as you edit the file.
+Removes redunant import statements of `react` in components.
```
Comment (pczern) [inline README.md:61]: typo `redundant`

File: src/pages/_app.tsx [hunk 6]
```diff
@@ -1,7 +1,19 @@
...
+
+import { BASE_URL } from '@/util/constants'
 import '@/styles/globals.css'
-import type { AppProps } from 'next/app'
 
-export default function App({ Component, pageProps }: AppProps) {
-  /* eslint-disable-next-line */
+function App({ Component, pageProps }: AppProps) {
+  /* eslint-disable react/jsx-props-no-spreading */
   return <Component {...pageProps} />
```
Comment (pczern) [inline src/pages/_app.tsx:10]: this was also done in Arnos PR

Comment (pczern) [reply src/pages/_app.tsx:10]: same way

File: src/util/constants.ts [hunk 7]
```diff
@@ -0,0 +1,3 @@
+// GraphQL endpoint
+/* eslint-disable import/prefer-default-export */ // remove this once more constants are added
+export const BASE_URL = process.env.BASE_URL || 'http://localhost:3000/graphql'
```
Comment (pczern) [inline src/util/constants.ts:3]: there is a `constants` folder, we shouldn't put in `util` folder

File: .eslintrc.js (hunk 1 above)
Comment (cnk1dev) [reply .eslintrc.js:34]: Screenshot 2023-05-14 at 01 01 34

@pczern

File: .eslintrc.js (hunk 2 above)
Comment (cnk1dev) [reply .eslintrc.js:22]: ok, we can just use Arno's version then when we merge

File: src/pages/_app.tsx (hunk 6 above)
Comment (cnk1dev) [reply src/pages/_app.tsx:10]: ok, should not be an issue when we merge then

File: .eslintrc.js (hunk 1 above)
Comment (pczern) [reply .eslintrc.js:34]: I think it is better to use 
```
 "globals": {
        "process": true
      }
```

Comment (pczern) [reply .eslintrc.js:34]: or take a look at this, if it's up to date https://stackoverflow.com/a/76081175

File: src/pages/_app.tsx [hunk 8]
```diff
@@ -1,7 +1,19 @@
...
 }
+
+export default withUrqlClient(
+  (ssrExchange) => ({
+    url: BASE_URL,
+    exchanges: [cacheExchange, ssrExchange, fetchExchange]
```
Comment (pczern) [inline src/pages/_app.tsx:16]: in what case do we need ssrExchange?

Comment (pczern) [reply src/pages/_app.tsx:16]: I think we could do a pretty basic non ssr setup at the moment, this might then even be a lot faster for page load

**Review by pczern — CHANGES_REQUESTED** [review]
Please change to globals process or import globals like in the stackoverflow link,
give further explanation about Codegen and how gitRepositories.tsx is generated

File: .eslintrc.js (hunk 1 above)
Comment (cnk1dev) [reply .eslintrc.js]: done

File: src/pages/_app.tsx (hunk 8 above)
Comment (cnk1dev) [reply src/pages/_app.tsx:16]: Are you sure CSR would be faster? Afaik initial page load is faster with SSR than CSR, but subsequent page loads are faster with CSR. Neither setup is very complex with urql though, so if we make the decision to use CSR we can easily change it. But then again, what are we using Next for?

Comment (cnk1dev) [reply src/pages/_app.tsx:16]: The codegen command is explained in the readme, all it does is create types and a custom hook for each query

Comment (cnk1dev) [reply src/pages/_app.tsx]: Removed ssrExchange

**Review by pczern — APPROVED** [review]
If `ts-node` is not needed please remove, otherwise please merge

**Pull request merged**